Build 2.4.x, bump 2.3.x, and drop patch number from tags

jeancochrane opened the pull request:
## Overview

Make a few small updates to builds, including:

* Build 2.4.x images
* Bump 2.3.x images to 2.3.3
* Drop the patch number from the image tags (e.g. 2.4.0 -> 2.4)
* Install `proj4` development files in Alpine images to make sure that GDAL gets compiled with PROJ4 support

Closes #2 and closes #3.

## Testing instructions

* Confirm that the Travis builds run successfully
* Pull the latest `master` commit from https://github.com/geotrellis/geotrellis-gdal
* In `geotrellis-gdal`, update `docker-compose.yml` to point to a test 2.4 image that I built:

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index ef5865b..4920c7a 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1,7 +1,7 @@
 version: '2.3'
 services:
   geotrellis-gdal:
-    image: quay.io/azavea/openjdk-gdal:2.3.2-jdk8-slim
+    image: quay.io/azavea/openjdk-gdal:2.4-jdk8-alpine-test
     environment:
       - BINTRAY_USER=$BINTRAY_USER
       - BINTRAY_PASS=$BINTRAY_PASS
```

* In `geotrellis-gdal`, run `./scripts/test` to run tests and confirm that they pass

Commit:
Build 2.4.x, bump 2.3.x, and drop patch from tags

* Build 2.4.x images
* Bump 2.3.x images to 2.3.3
* Drop the patch number from the image tags (e.g. 2.4.0 -> 2.4)

Commit:
Add proj4 development package to Alpine images

Previously, Alpine images were not installing proj4-dev, which
would cause GDAL to get compiled without PROJ4 support. Install
proj4-dev in Alpine images to fix this compilation error.

jeancochrane commented on Dockerfile-alpine.template:
```diff
@@ -49,13 +50,11 @@ RUN set -ex \
     && ./configure --with-python --with-curl --with-java=${JAVA_HOME} --with-openjpeg=$ROOTDIR \
     && make -j3 && make -j3 install \
-    # Compile Python and Java bindings for GDAL
+    # Compile Python bindings for GDAL
```
I made this comment before I fully understood what removing lines 57-8 meant -- we are in fact still installing Java bindings in this block of code, we just aren't moving the shared objects over manually. I can revert this change if it's confusing.

jeancochrane commented on scripts/cibuild:
```diff
@@ -23,8 +23,11 @@ if [ "${BASH_SOURCE[0]}" = "${0}" ]; then
             "Dockerfile-$VARIANT.template" > Dockerfile
 
+        # Strip the patch number off of GDAL_VERSION for tagging
+        GDAL_MAJOR_VERSION="${GDAL_VERSION:0:3}"
```
There may be a more centralized way of doing this, as opposed to duplicating the string manipulation in both `cibuild`, `cipublish`, and `test`. The first thing that came to mind was requiring that the major version be its own environment variable, but I didn't like the redundancy there. Other ideas are welcome.

hectcastro replied:
This works, but could break down in the event that the MINOR version is double digits. Might be able to be more precise with `cut`:

```bash
echo "${GDAL_VERSION}" | cut -d"." -f1,2
```

jeancochrane replied:
Much better. Slightly ashamed that I decided to go with a spotty string manipulation after [getting burned by a spotty string manipulation yesterday](https://github.com/raster-foundry/load-testing/pull/2).

Commit:
Use GDAL_MAJOR_VERSION when publishing packages

Make sure to reference the GDAL_MAJOR_VERSION env variable when
running docker push, so that the correct images get uploaded.

hectcastro reviewed (approved):
Looks good. I was able to get the `geotrellis-gdal` tests running after applying the following patch:

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index ef5865b..361969d 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -1,7 +1,7 @@
 version: '2.3'
 services:
   geotrellis-gdal:
-    image: quay.io/azavea/openjdk-gdal:2.3.2-jdk8-slim
+    image: quay.io/azavea/openjdk-gdal:2.4-jdk8-alpine-test
     environment:
       - BINTRAY_USER=$BINTRAY_USER
       - BINTRAY_PASS=$BINTRAY_PASS
@@ -12,4 +12,4 @@ services:
       - $HOME/.sbt:/root/.sbt
       - $HOME/.coursier:/root/.coursier
     working_dir: /opt/geotrellis-gdal
-    entrypoint: ./sbt "++$SCALA_VERSION"
+    entrypoint: ./sbt "++2.11.12"
```

I opened https://github.com/geotrellis/geotrellis-gdal/issues/51 to make a more permanent fix.

Commit:
Use cut instead of substring to strip GDAL patch number

In order to remove the GDAL version patch number from image tag strings,
use a safer 'cut'-based strategy instead of direct substring manipulation
in Bash.
